File: 0337_House Robber III/HouseRobberIIIMemorization.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    private HashMap<TreeNode,Integer> map = new HashMap();
    public int rob(TreeNode root) {
        if(root == null) return 0;
        if(map.containsKey(root)) return map.get(root);

        int val = root.val;
        int ll = root.left != null ? rob(root.left.left) : 0;
        int lr = root.left != null ? rob(root.left.right) : 0;
        int rl = root.right != null ? rob(root.right.left) : 0;
        int rr = root.right != null ? rob(root.right.right) : 0;

        int max = Math.max(val + ll + lr + rl + rr, rob(root.left) + rob(root.right));
        map.put(root, max);
        return max;
    }
}

File: 0337_House Robber III/HouseRobberIIIMemorization.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    public int Rob(TreeNode root) {
        return Rob(root, new Dictionary<TreeNode, int>());
    }
    
    private int Rob(TreeNode node, Dictionary<TreeNode, int> dict)
    {
        if(node == null) return 0;
        if(dict.ContainsKey(node)) return dict[node];
        var v = node.val;
        var ll = node.left != null ? Rob(node.left.left, dict) : 0;
        var lr = node.left != null ? Rob(node.left.right, dict) : 0;
        var rl = node.right != null ? Rob(node.right.left, dict) : 0;
        var rr = node.right != null ? Rob(node.right.right, dict) : 0;
        var k = Math.Max(v + ll + lr + rl + rr, Rob(node.left, dict) + Rob(node.right, dict));
        dict.Add(node, k);
        return k;
    }
}